File: src/readers/image/decoder.ts
/** Image Decoder Options */
export interface DecodeOptions {
  x?: number;
  y?: number;
  width?: number;
  height?: number;
  /** The modulo to use. For example you may have a 514x514 image, but you want to use 512x512 [Default=1] */
  modulo?: number;
}

/**
 * Image decoder
 * @param buffer - the input buffer
 * @param options - user defined options
 * @returns - the decoded buffer
 */
export async function imageDecoder(
  buffer: ArrayBufferLike | Uint8Array | Buffer,
  options?: DecodeOptions,
): Promise<ImageData> {
  const blob = new Blob([buffer as ArrayBuffer]); // e.g. { type: 'image/png' }
  const imageBitmap = await createImageBitmap(blob);
  let { x = 0, y = 0, width = imageBitmap.width, height = imageBitmap.height } = options ?? {};
  const { modulo = 1 } = options ?? {};
  // If modulo exists, shift x and y by modulo's diff
  const diff = imageBitmap.width % modulo;
  if (diff !== 0) {
    x = y = Math.min(diff / 2);
    width = height = imageBitmap.width - diff;
  }
  // Create OffscreenCanvas and draw
  const canvas = new OffscreenCanvas(imageBitmap.width, imageBitmap.height);
  const ctx = canvas.getContext('2d');
  if (ctx === null) throw new Error('Could not get 2d context');
  ctx.drawImage(imageBitmap, 0, 0);

  return ctx.getImageData(x, y, width, height);
}

/**
 * Image decoder
 * @param buffer - the input buffer
 * @param options - user defined options
 * @returns - the decoded buffer
 */
export async function imageDecoderBuffer(
  buffer: ArrayBufferLike | Uint8Array | Buffer,
  options?: DecodeOptions,
): Promise<ArrayBufferLike> {
  const imageData = await imageDecoder(buffer, options);
  return imageData.data.buffer;
}

File: rust/parsers/image/decoder.rs
#[cfg(feature = "std")]
use super::ImageData;
#[cfg(feature = "std")]
use crate::parsers::Buffer;
#[cfg(feature = "std")]
use image::ImageReader;
#[cfg(feature = "std")]
use std::io::Cursor;

/// Image Decoder Options
#[derive(Debug, Default, Clone)]
pub struct DecodeOptions {
    /// The x position to start at
    pub x: Option<u32>,
    /// The y position to start at
    pub y: Option<u32>,
    /// The width to read in
    pub width: Option<u32>,
    /// The height to read in
    pub height: Option<u32>,
    /// The modulo to use. For example you may have a 514x514 image, but you want to use 512x512 [Default=1]
    pub modulo: Option<u32>,
}

/// Decode any image
///
/// ## Parameters
/// - `buffer`: the input buffer
/// - `options`: the decode options
///
/// ## Returns
/// the raw decoded buffer
#[cfg(feature = "std")]
pub fn image_decoder(
    buffer: &Buffer,
    options: Option<DecodeOptions>,
) -> Result<ImageData, image::ImageError> {
    use image::GenericImageView;

    let img = ImageReader::new(Cursor::new(buffer.buf())).with_guessed_format()?.decode()?;

    let (img_width, img_height) = img.dimensions();
    let opts = options.unwrap_or_default();

    let modulo = opts.modulo.unwrap_or(1);
    let diff = img_width % modulo;

    let (x, y, width, height) = if diff != 0 {
        let shift = diff / 2;
        (shift, shift, img_width - diff, img_width - diff)
    } else {
        (
            opts.x.unwrap_or(0),
            opts.y.unwrap_or(0),
            opts.width.unwrap_or(img_width),
            opts.height.unwrap_or(img_height),
        )
    };

    let sub_image = img.view(x, y, width, height).to_image();

    Ok(ImageData {
        width: width as usize,
        height: height as usize,
        data: sub_image.into_raw().into(),
    })
}

/// Image decoder
///
/// ## Parameters
/// - `buffer`: the input buffer
/// - `options`: user defined options
///
/// ## Returns
/// the decoded buffer
#[cfg(feature = "std")]
pub fn image_decoder_buffer(buffer: &Buffer, options: Option<DecodeOptions>) -> Buffer {
    image_decoder(buffer, options).unwrap().data
}
